### PR TITLE
Remove the stale effecters and sensors after the CM

### PR DESCRIPTION
**Pull request by Pavithrab7**
This commit removes the duplicate sensor and effecter PDRs
that were present in our repo after the CM/hotplug.

tested the CM operation scenarios.

Fixes:SW548661

Signed-off-by: Pavithra Barithaya <pavithra.b@ibm.com>

### CODE REVIEW

File: common/utils.cpp [hunk 1]
```diff
@@ -685,5 +685,185 @@ std::string getCurrentSystemTime()
     return ss.str();
 }
 
+std::vector<std::vector<uint8_t>> findAllStateEffecterPDRs(uint8_t /*tid*/,
```
Comment (manojkiraneda) [inline common/utils.cpp:688]: getStateEffecterPDRsByType makes more sense right ?

Comment (manojkiraneda) [reply common/utils.cpp:688]: std::vector<std::vector<uint8_t>> -> use pldm types to make it implicit to understand what you are trying to return.

File: common/utils.cpp [hunk 2]
```diff
@@ -685,5 +685,185 @@ std::string getCurrentSystemTime()
...
+
+    try
+    {
+        do
+        {
+            record = pldm_pdr_find_record_by_type(repo, PLDM_STATE_EFFECTER_PDR,
+                                                  record, &outData, &size);
+            if (record)
+            {
+                auto pdr = reinterpret_cast<pldm_state_effecter_pdr*>(outData);
```
Comment (manojkiraneda) [inline common/utils.cpp:705]: nullptr check missing for pdr

File: common/utils.cpp [hunk 3]
```diff
@@ -685,5 +685,185 @@ std::string getCurrentSystemTime()
...
+                                             sizeof(possibleStateSize);
+                }
+            }
+
+        } while (record);
+    }
+    catch (const std::exception& e)
+    {
+        std::cerr << " Failed to obtain a record. ERROR =" << e.what()
+                  << std::endl;
```
Comment (manojkiraneda) [inline common/utils.cpp:736]: I am not sure why this in a try-catch block ? does the underlying code generate an exception ? can you explain what kind of exceptions do you foresee this catching ?

File: common/utils.cpp [hunk 4]
```diff
@@ -685,5 +685,185 @@ std::string getCurrentSystemTime()
...
+    }
+
+    return pdrs;
+}
+
+std::vector<std::vector<uint8_t>> findAllStateSensorPDRs(uint8_t /*tid*/,
```
Comment (manojkiraneda) [inline common/utils.cpp:742]: same comments here as well ,  use the types in the return value to make things implicitly understandable.

change the function name

File: common/utils.cpp [hunk 5]
```diff
@@ -685,5 +685,185 @@ std::string getCurrentSystemTime()
...
+    std::vector<std::vector<uint8_t>> pdrs;
+    try
+    {
+        do
+        {
+            record = pldm_pdr_find_record_by_type(repo, PLDM_STATE_SENSOR_PDR,
+                                                  record, &outData, &size);
+            if (record)
+            {
+                auto pdr = reinterpret_cast<pldm_state_sensor_pdr*>(outData);
```
Comment (manojkiraneda) [inline common/utils.cpp:758]: nullptr check missing.

File: common/utils.cpp [hunk 6]
```diff
@@ -685,5 +685,185 @@ std::string getCurrentSystemTime()
...
+                }
+            }
+
+        } while (record);
+    }
+    catch (const std::exception& e)
+    {
+        std::cerr << " Failed to obtain a record. ERROR =" << e.what()
+                  << std::endl;
+    }
```
Comment (manojkiraneda) [inline common/utils.cpp:790]: same comments here as well.

File: common/utils.cpp [hunk 7]
```diff
@@ -685,5 +685,185 @@ std::string getCurrentSystemTime()
...
+
+std::vector<uint16_t> findEffecterIds(const pldm_pdr* pdrRepo, uint8_t tid,
+                                      uint16_t entityType,
+                                      uint16_t entityInstance,
+                                      uint16_t containerId)
+{
+    std::vector<uint16_t> effecterIDs;
+
+    auto pdrs = findAllStateEffecterPDRs(tid, entityType, pdrRepo);
+    for (auto pdr : pdrs)
```
Comment (manojkiraneda) [inline common/utils.cpp:803]: unnecessary copy - const auto&

File: common/utils.cpp [hunk 8]
```diff
@@ -685,5 +685,185 @@ std::string getCurrentSystemTime()
...
+    {
+        auto effecterPdr =
```
Comment (manojkiraneda) [inline common/utils.cpp:805]: nullptr check missing

File: common/utils.cpp [hunk 9]
```diff
@@ -685,5 +685,185 @@ std::string getCurrentSystemTime()
...
+    catch (const std::exception& e)
+    {
+        std::cerr << " Failed to obtain a record. ERROR =" << e.what()
+                  << std::endl;
+    }
+
+    return pdrs;
+}
+
+std::vector<uint16_t> findEffecterIds(const pldm_pdr* pdrRepo, uint8_t tid,
```
Comment (manojkiraneda) [inline common/utils.cpp:795]: use typedefinitions in the return vaue.

File: common/utils.cpp [hunk 10]
```diff
@@ -685,5 +685,185 @@ std::string getCurrentSystemTime()
...
+std::vector<uint16_t> findSensorIds(const pldm_pdr* pdrRepo, uint8_t tid,
+                                    uint16_t entityType,
+                                    uint16_t entityInstance,
+                                    uint16_t containerId)
+{
+
+    std::vector<uint16_t> sensorIDs;
+
+    auto pdrs = findAllStateSensorPDRs(tid, entityType, pdrRepo);
+    for (auto pdr : pdrs)
```
Comment (manojkiraneda) [inline common/utils.cpp:841]: same comments here

File: common/utils.cpp [hunk 11]
```diff
@@ -685,5 +685,185 @@ std::string getCurrentSystemTime()
...
+            }
+            possible_states_start += possibleStateSize +
+                                     sizeof(possibleStates->state_set_id) +
+                                     sizeof(possibleStateSize);
+        }
+    }
+    return effecterIDs;
+}
+
+std::vector<uint16_t> findSensorIds(const pldm_pdr* pdrRepo, uint8_t tid,
```
Comment (manojkiraneda) [inline common/utils.cpp:832]: same comments here

File: common/utils.cpp [hunk 12]
```diff
@@ -685,5 +685,185 @@ std::string getCurrentSystemTime()
...
+                                    uint16_t entityInstance,
+                                    uint16_t containerId)
+{
+
+    std::vector<uint16_t> sensorIDs;
+
+    auto pdrs = findAllStateSensorPDRs(tid, entityType, pdrRepo);
+    for (auto pdr : pdrs)
+    {
+        auto sensorPdr = reinterpret_cast<pldm_state_sensor_pdr*>(pdr.data());
```
Comment (manojkiraneda) [inline common/utils.cpp:843]: nullptr check missing

File: libpldmresponder/fru.cpp [hunk 13]
```diff
@@ -437,8 +438,36 @@ void FruImpl::removeIndividualFRU(const std::string& fruObjPath)
...
         std::move(std::vector<ChangeEntry>(1, deleteRecordHdl)),
-        std::move(std::vector<uint8_t>(1, PLDM_RECORDS_DELETED))); // need to
-    //  send both remote and local records. Phyp keeps track of bmc only records
+        std::move(std::vector<uint8_t>(1, PLDM_RECORDS_DELETED)));
+
+    std::vector<uint16_t> effecterIDs = findEffecterIds(
+        pdrRepo, 0 /*tid*/, removeEntity.entity_type,
+        removeEntity.entity_instance_num, removeEntity.entity_container_id);
+
+    for (auto& ids : effecterIDs)
```
Comment (manojkiraneda) [inline libpldmresponder/fru.cpp:447]: const auto&

File: libpldmresponder/fru.cpp [hunk 14]
```diff
@@ -437,8 +438,36 @@ void FruImpl::removeIndividualFRU(const std::string& fruObjPath)
...
+    {
+        auto delEffecterHdl = pldm_delete_by_effecter_id(pdrRepo, ids, false);
```
Comment (manojkiraneda) [inline libpldmresponder/fru.cpp:449]: check if the handle is 0 before sending the event ?

File: common/utils.cpp (hunk 8 above)
Comment (Pavithrab7) [reply common/utils.cpp:805]: done

File: common/utils.cpp (hunk 7 above)
Comment (Pavithrab7) [reply common/utils.cpp:803]: done

File: common/utils.cpp (hunk 6 above)
Comment (Pavithrab7) [reply common/utils.cpp:790]: done

File: common/utils.cpp (hunk 5 above)
Comment (Pavithrab7) [reply common/utils.cpp:758]: done

File: common/utils.cpp (hunk 4 above)
Comment (Pavithrab7) [reply common/utils.cpp:742]: done

File: common/utils.cpp (hunk 3 above)
Comment (Pavithrab7) [reply common/utils.cpp:736]: I didn't have a null pointer check, I presume the null pointer check will handle the error conditions. Removed the try and catch block.

File: common/utils.cpp (hunk 2 above)
Comment (Pavithrab7) [reply common/utils.cpp:705]: done

File: common/utils.cpp (hunk 9 above)
Comment (Pavithrab7) [reply common/utils.cpp:795]: done

File: common/utils.cpp (hunk 10 above)
Comment (Pavithrab7) [reply common/utils.cpp:841]: done

File: common/utils.cpp (hunk 11 above)
Comment (Pavithrab7) [reply common/utils.cpp:832]: done

File: common/utils.cpp (hunk 12 above)
Comment (Pavithrab7) [reply common/utils.cpp:843]: done

File: common/utils.cpp (hunk 1 above)
Comment (Pavithrab7) [reply common/utils.cpp:688]: done

File: libpldmresponder/fru.cpp [hunk 15]
```diff
@@ -437,8 +438,36 @@ void FruImpl::removeIndividualFRU(const std::string& fruObjPath)
...
+        pdrRepo, 0 /*tid*/, removeEntity.entity_type,
+        removeEntity.entity_instance_num, removeEntity.entity_container_id);
+
+    for (auto& ids : sensorIDs)
+    {
+        auto delSensorHdl = pldm_delete_by_sensor_id(pdrRepo, ids, false);
+        sendPDRRepositoryChgEventbyPDRHandles(
+            std::move(std::vector<ChangeEntry>(1, delSensorHdl)),
+            std::move(std::vector<uint8_t>(1, PLDM_RECORDS_DELETED)));
+        sensorDbusObjMaps.erase(ids);
```
[the remark this answers is not in the document]
Comment (Pavithrab7) [reply libpldmresponder/fru.cpp:466]: done

File: libpldmresponder/fru.cpp [hunk 16]
```diff
@@ -437,8 +438,36 @@ void FruImpl::removeIndividualFRU(const std::string& fruObjPath)
...
+
+    std::vector<uint16_t> sensorIDs = findSensorIds(
+        pdrRepo, 0 /*tid*/, removeEntity.entity_type,
+        removeEntity.entity_instance_num, removeEntity.entity_container_id);
+
+    for (auto& ids : sensorIDs)
+    {
+        auto delSensorHdl = pldm_delete_by_sensor_id(pdrRepo, ids, false);
+        sendPDRRepositoryChgEventbyPDRHandles(
+            std::move(std::vector<ChangeEntry>(1, delSensorHdl)),
```
[the remark this answers is not in the document]
Comment (Pavithrab7) [reply libpldmresponder/fru.cpp:464]: done

File: libpldmresponder/fru.cpp (hunk 14 above)
Comment (Pavithrab7) [reply libpldmresponder/fru.cpp:449]: done

File: libpldmresponder/fru.cpp (hunk 13 above)
Comment (Pavithrab7) [reply libpldmresponder/fru.cpp:447]: done

**Review by manojkiraneda — APPROVED** [review]

**Review by sampmisr — APPROVED** [review]

**Pull request merged**